File: src/main/java/com/refinedmods/refinedstorage/tile/config/IType.java
package com.refinedmods.refinedstorage.tile.config;

import com.refinedmods.refinedstorage.api.network.node.INetworkNodeProxy;
import com.refinedmods.refinedstorage.inventory.fluid.FluidInventory;
import com.refinedmods.refinedstorage.tile.data.TileDataParameter;
import com.refinedmods.refinedstorage.tile.data.TileDataParameterClientListener;
import net.minecraft.network.syncher.EntityDataSerializers;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraftforge.items.IItemHandlerModifiable;

import javax.annotation.Nullable;

public interface IType {
    int ITEMS = 0;
    int FLUIDS = 1;

    static <T extends BlockEntity & INetworkNodeProxy<?>> TileDataParameter<Integer, T> createParameter(@Nullable TileDataParameterClientListener<Integer> clientListener) {
        return new TileDataParameter<>(EntityDataSerializers.INT, ITEMS, t -> ((IType) t.getNode()).getType(), (t, v) -> {
            if (v == IType.ITEMS || v == IType.FLUIDS) {
                ((IType) t.getNode()).setType(v);
            }
        }, clientListener);
    }

    static <T extends BlockEntity & INetworkNodeProxy<?>> TileDataParameter<Integer, T> createParameter() {
        return createParameter(null);
    }

    int getType();

    void setType(int type);

    IItemHandlerModifiable getItemFilters();

    FluidInventory getFluidFilters();
}

File: src/main/kotlin/com/refinedmods/refinedstorage/tile/config/IType.kt
package com.refinedmods.refinedstorage.tile.config

import com.refinedmods.refinedstorage.api.network.node.INetworkNodeProxy
import com.refinedmods.refinedstorage.tile.data.TileDataParameter
import com.refinedmods.refinedstorage.tile.data.TileDataParameterClientListener
import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.data.TrackedData
import net.minecraft.entity.data.TrackedDataHandlerRegistry
import net.minecraft.inventory.Inventory
import reborncore.common.util.Tank
import java.util.function.BiConsumer
import java.util.function.Function

interface IType {
    var type: Int
    val itemFilters: Inventory?
    val fluidFilters: Tank?

    companion object {
        fun <T> createParameter(clientListener: TileDataParameterClientListener<Int?>?): TileDataParameter<Int, T> where T : BlockEntity?, T : INetworkNodeProxy<*>? {
            return TileDataParameter<Int, T>(
                    ITEMS,
                    TrackedDataHandlerRegistry.INTEGER,
                    Function { t: T? -> (t!!.node as IType).type },
                    BiConsumer { t: T?, v: Int? ->
                        if (v == ITEMS || v == FLUIDS) {
                            (t!!.node as IType).type = v
                        }
                    },
                    clientListener
            )
        }

        fun <T> createParameter(): TileDataParameter<Int, T> where T : BlockEntity?, T : INetworkNodeProxy<*>? {
            return createParameter(null)
        }

        const val ITEMS = 0
        const val FLUIDS = 1
    }
}